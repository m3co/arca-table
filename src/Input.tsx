import * as React from 'react';

interface Props {
  type?: string;
  value: string | number | boolean;
  onBlur?: (currentValue: string | number | boolean) => void;
  onEnter?: (currentValue: string | number | boolean) => void;
  onEsc?: () => void;
}

interface DefaultProps {
  type: string,
}

interface State {
  value: string | number | boolean;
}

export class Input
  extends React.Component<Props>
{
  public state: State = {
    value: '',
  }

  public static defaultProps: DefaultProps = {
    type: 'text',
  }

  public constructor(props: Props) {
    super(props)
    this.state.value = this.props.value;
    if (this.props.type?.toLowerCase() === 'date') {
      this.state.value = this.props.value ?
        this.toDate(this.dateToArray(new Date(this.props.value.toString()))) :
        '';
    }
    if (this.props.type?.toLowerCase() === 'datetime-local') {
      this.state.value = this.props.value ?
        this.toDateTime(this.dateToArray(new Date(this.props.value.toString()))) :
        '';
    }
  }

  private dateToArray(d: Date) {
    return d.toLocaleString(undefined, {
      year: 'numeric',
      month: '2-digit',
      day: '2-digit',
      hour: '2-digit',
      minute: '2-digit',
      second: '2-digit',
      hour12: false,
    }).match(/(\d+)/g)
  }

  private toDate(d: RegExpMatchArray | null) {
    return d === null ? '' : `${d[2]}-${d[0]}-${d[1]}`;
  }

  private toDateTime(d: RegExpMatchArray | null) {
    return d === null ? '' : `${d[2]}-${d[0]}-${d[1]}T${d[3]}:${d[4]}:${d[5]}`;
  }

  private onChange = (event: React.ChangeEvent<HTMLInputElement>): void => {
    const { type } = this.props;
    if (type) {
      switch (type.toLowerCase()) {
        case 'number':
          return this.setState({ value: Number(event.target.value) });
        case 'checkbox':
          return this.setState({ value: Boolean(event.target.checked) });
        default:
          break;
      }
    }
    this.setState({ value: event.target.value });
  }

  private onKeyUp = (event: React.KeyboardEvent<HTMLInputElement>): void => {
    const extra = {
      altKey: !!event.altKey,
      ctrlKey: !!event.ctrlKey,
      metaKey: !!event.metaKey,
      shiftKey: !!event.shiftKey,
    };
    if (extra.altKey || extra.ctrlKey || extra.metaKey || extra.shiftKey) {
      return;
    }
    if (event.key == 'Enter' && this.props.onEnter) {
      this.props.onEnter(this.state.value);
    }
    if (event.key == 'Escape' && this.props.onEsc) {
      this.props.onEsc();
    }
  }

  private onBlur = (): void => {
    if (this.props.onBlur) this.props.onBlur(this.state.value);
  }

  public render(): JSX.Element {
    const { value } = this.state;
    const { type } = this.props;
    return (
      <input
        type={type}
        autoFocus={true}
        onChange={this.onChange}
        onKeyUp={this.onKeyUp}
        onBlur={this.onBlur}
        value={value.toString()}></input>
    );
  }
}
