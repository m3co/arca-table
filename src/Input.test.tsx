
import * as React from 'react';
import { mount } from 'enzyme';

import { Input } from './Input';

test('Input normally', (): void => {
  const el = (<Input value={'value A'}/>);
  const wrapper = mount(el);
  expect(wrapper).toMatchSnapshot();
});

test('Input changes its value', (): void => {
  const el = (<Input value={'value A'}/>);
  const wrapper = mount(el);
  wrapper.find('input').simulate('change', { target: { value: 'value B' } });
  expect(wrapper).toMatchSnapshot();
});

test('Input changes its numeric value', (): void => {
  const el = (<Input value={10} type="number" />);
  const wrapper = mount(el);
  wrapper.find('input').simulate('change', { target: { value: 20 } });
  expect(wrapper).toMatchSnapshot();
});


test('Input fires onBlur', (): void => {
  const value = 'value A';
  let onBlurReached = false;
  const onBlur = (currentValue: string | number | boolean): void => {
    expect(currentValue).toBe(value);
    onBlurReached = true;
  }
  const el = (<Input value={value} onBlur={onBlur}/>);
  const wrapper = mount(el);
  wrapper.find('input').simulate('blur');
  expect(onBlurReached).toBe(true);
});

test('Input ignores onEnter if altKey', (): void => {
  const value = 'value A';
  const onEnter = (currentValue: string | number | boolean): void => {
    expect(currentValue).toBe(value);
    fail();
  }
  const el = (<Input value={value} onEnter={onEnter}/>);
  const wrapper = mount(el);
  wrapper.find('input').simulate('keyup', { key: 'Enter', altKey: true });
});

test('Input fires onEnter', (): void => {
  const value = 'value A';
  let onEnterReached = false;
  const onEnter = (currentValue: string | number | boolean): void => {
    expect(currentValue).toBe(value);
    onEnterReached = true;
  }
  const el = (<Input value={value} onEnter={onEnter}/>);
  const wrapper = mount(el);
  wrapper.find('input').simulate('keyup', { key: 'Enter' });
  expect(onEnterReached).toBe(true);
});

test('Input fires onEsc', (): void => {
  const value = 'value A';
  let onEscReached = false;
  const onEsc = (): void => {
    onEscReached = true;
  }
  const el = (<Input value={value} onEsc={onEsc}/>);
  const wrapper = mount(el);
  wrapper.find('input').simulate('keyup', { key: 'Escape' });
  expect(onEscReached).toBe(true);
});
