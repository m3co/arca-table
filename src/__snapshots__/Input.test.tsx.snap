// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input changes its value 1`] = `
<Input
  value="value A"
>
  <input
    autoFocus={true}
    onBlur={[Function]}
    onChange={[Function]}
    onKeyUp={[Function]}
    value="value B"
  />
</Input>
`;

exports[`Input normally 1`] = `
<Input
  value="value A"
>
  <input
    autoFocus={true}
    onBlur={[Function]}
    onChange={[Function]}
    onKeyUp={[Function]}
    value="value A"
  />
</Input>
`;
